src/datalink/libpcap.rs

```diff
@@ -0,0 +1,194 @@
+                      func: &mut FnMut(MutableEthernetPacket))
+                      -> Option<io::Result<()>> {
+        for _ in 0..num_packets {
+            let mut data = vec![0; packet_size];
+            {
+                let eh = MutableEthernetPacket::new(&mut data).unwrap();
+                func(eh);
+            }
+            let mut cap = self.capture.lock().unwrap();
+            if let Err(e) = cap.sendpacket(&data) {
+                return Some(Err(io::Error::new(io::ErrorKind::Other, e)))
+            }
+        }
+        Some(Ok(()))
+    }
+
+    #[inline]
+    fn send_to(&mut self,
+               packet: &EthernetPacket,
+               _dst: Option<NetworkInterface>) -> Option<io::Result<()>> {
+        let mut cap = self.capture.lock().unwrap();
+        Some(match cap.sendpacket(packet.packet()) {
+            Ok(()) => Ok(()),
+            Err(e) => Err(io::Error::new(io::ErrorKind::Other, e)),
+        })
+    }
+}
+
+struct InvalidDataLinkSenderImpl {}
+
+impl EthernetDataLinkSender for InvalidDataLinkSenderImpl {
+    #[inline]
+    fn build_and_send(&mut self,
+                      _num_packets: usize,
+                      _packet_size: usize,
+                      _func: &mut FnMut(MutableEthernetPacket))
+                      -> Option<io::Result<()>> {
+        None
+    }
+
+    #[inline]
+    fn send_to(&mut self,
+               _packet: &EthernetPacket,
+               _dst: Option<NetworkInterface>) -> Option<io::Result<()>> {
+        None
+    }
+}
+
+struct DataLinkReceiverImpl<T: Activated + Send + Sync> {
+    capture: Arc<Mutex<pcap::Capture<T>>>,
+    read_buffer: Vec<u8>,
+}
+
+impl <T: Activated + Send + Sync> EthernetDataLinkReceiver for DataLinkReceiverImpl<T> {
+    fn iter<'a>(&'a mut self) -> Box<EthernetDataLinkChannelIterator + 'a> {
+        Box::new(DataLinkChannelIteratorImpl { pc: self })
+    }
+}
+
+struct DataLinkChannelIteratorImpl<'a, T: Activated + Send + Sync + 'a> {
+    pc: &'a mut DataLinkReceiverImpl<T>,
+}
+
+impl<'a, T: Activated + Send + Sync> EthernetDataLinkChannelIterator<'a> for DataLinkChannelIteratorImpl<'a, T> {
+    fn next(&mut self) -> io::Result<EthernetPacket> {
+        let mut cap = self.pc.capture.lock().unwrap();
+        match cap.next() {
+            Ok(pkt) => {
+                self.pc.read_buffer.truncate(0);
+                self.pc.read_buffer.extend(pkt.data);
+            },
+            Err(e) => return Err(io::Error::new(io::ErrorKind::Other, e)),
+        };
+        Ok(EthernetPacket::new(&self.pc.read_buffer).unwrap())
+    }
+}
+
+/// Get a list of available network interfaces for the current machine.
+pub fn interfaces() -> Vec<NetworkInterface> {
+    if let Ok(devices) = pcap::Device::list() {
+        devices.iter().enumerate().map(|(i, dev)| {
+            NetworkInterface {
+                name: dev.name.clone(),
+                index: i as u32,
+                mac: None,
+                ips: None,
```

Comment:
They have an open issue for it, might be worth helping them out: https://github.com/ebfull/pcap/issues/13